gluten-core/src/main/scala/io/glutenproject/execution/BroadcastNestedLoopJoinExecTransformer.scala

```diff
@@ -0,0 +1,206 @@
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.glutenproject.execution
+
+import io.glutenproject.backendsapi.BackendsApiManager
+import io.glutenproject.expression.ExpressionConverter
+import io.glutenproject.extension.ValidationResult
+import io.glutenproject.metrics.MetricsUpdater
+import io.glutenproject.substrait.SubstraitContext
+import io.glutenproject.substrait.rel.RelBuilder
+import io.glutenproject.utils.SubstraitUtil
+
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.catalyst.expressions.{Attribute, Expression}
+import org.apache.spark.sql.catalyst.optimizer.{BuildLeft, BuildRight, BuildSide}
+import org.apache.spark.sql.catalyst.plans.{InnerLike, JoinType, LeftOuter, RightOuter}
+import org.apache.spark.sql.catalyst.plans.physical.Partitioning
+import org.apache.spark.sql.execution.{SparkPlan, SQLExecution}
+import org.apache.spark.sql.execution.joins.BaseJoinExec
+import org.apache.spark.sql.execution.metric.SQLMetric
+import org.apache.spark.sql.vectorized.ColumnarBatch
+
+import io.substrait.proto.CrossRel
+
+abstract class BroadcastNestedLoopJoinExecTransformer(
+    left: SparkPlan,
+    right: SparkPlan,
+    buildSide: BuildSide,
+    joinType: JoinType,
+    condition: Option[Expression])
+  extends BaseJoinExec
+  with TransformSupport {
+
+  def joinBuildSide: BuildSide = buildSide
+
+  override def leftKeys: Seq[Expression] = Nil
+  override def rightKeys: Seq[Expression] = Nil
+
+  private lazy val substraitJoinType: CrossRel.JoinType =
+    SubstraitUtil.toCrossRelSubstrait(joinType)
+
+  private lazy val buildTableId: String = "BuildTable-" + buildPlan.id
+
+  // Hint substrait to switch the left and right,
+  // since we assume always build right side in substrait.
+  private lazy val needSwitchChildren: Boolean = buildSide match {
+    case BuildLeft => true
+    case BuildRight => false
+  }
+
+  protected lazy val (buildPlan, streamedPlan) = if (needSwitchChildren) {
+    (left, right)
+  } else {
+    (right, left)
+  }
+
+  override def columnarInputRDDs: Seq[RDD[ColumnarBatch]] = {
+    val streamedRDD = getColumnarInputRDDs(streamedPlan)
+    val broadcastRDD = {
+      val executionId = sparkContext.getLocalProperty(SQLExecution.EXECUTION_ID_KEY)
+      BackendsApiManager.getBroadcastApiInstance
+        .collectExecutionBroadcastHashTableId(executionId, buildTableId)
```

Comment:
@zhztheplayer can I rename methods under `BroadcastApi` to make it more generic? Since now this is being used for both the modes i.e. `hashedBroadcastMode` & `IdentityBroadcastMode`. These method names containing `Hashtable` might confuse other developers who may read this file later.

Rename `collectExecutionBroadcastHashTableId` => `collectExecutionBroadcastTableId`
Rename `cleanExecutionBroadcastHashtable`=> `cleanExecutionBroadcastTable`

Reply:
> Rename collectExecutionBroadcastHashTableId => collectExecutionBroadcastTableId
Rename cleanExecutionBroadcastHashtable=> cleanExecutionBroadcastTable

Sounds fair to me. Thanks.